src/main.cpp

```diff
@@ -593,6 +593,9 @@ CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& loc
             CBlockIndex* pindex = (*mi).second;
             if (chain.Contains(pindex))
                 return pindex;
+            if (pindex->GetAncestor(chain.Height()) == chain.Tip()) {
```

Comment:
Maybe this should only do the `GetAncestor` check once, on the first entry in the block locator?